### ISSUE TITLE
Hello

### ISSUE THREAD

**Issue by nwulkow** [PhilippsCat/MarkovChains#1]
Hi, this is were we can communicate

**Comment by nbbl** [comment]
Hey everyone, I added a [wikipage](https://github.com/PhilippsCat/MarkovChains/wiki/Software-Development) about my proposal concerning branches. Constructive critic, contribution and requests are welcome :)

**Comment by c1887** [comment]
thank you very much for your proposal. I think it's a good idea using branches, otherwise we may get into unnecessary trouble. I still don't really understand whether the task is going to be more specified or if we just go on and implement basically everything we learn about in lecture.

**Comment by nbbl** [comment]
I added a [wiki page](https://github.com/PhilippsCat/MarkovChains/wiki/Task-assignment) where we can track what everyone is up to.

**Comment by PhilippsCat** [comment]
i wrote a function for hitting probabilities and comittors, but forgot how to upload something -.-

**Comment by PhilippsCat** [comment]
maybe somebody can explain it again?
via skype would be great because of "screen streaming" (is that the correct english word?)

**Comment by c1887** [comment]
I think Johannes's wikipage (linked above) describes it quite well?

I modified the MSM.py in order to make it work properly, as there were some issues when moving from the ipython notebook, where Niklas used "%pylab inline" to the standard .py format, which caused the class to not work properly. I've pushed my changes to my branch. Maybe one of you can review it and merge it to master if it's okay?

**Comment by nbbl** [comment]
@c1887 I looked through your code. It seemed fine, except for the [docstrings](http://en.wikipedia.org/wiki/Docstring#Python) which were in the wrong places. I solved this problem in my branch (see commit cda70b67).

Apart from that I implemented the basic checks that we discussed (see commit f674b76a), but I did not yet write any tests as it seems quite tedious to come up with lots of transition matrices ourselves. Maybe we can ask in the next exercise session if they can provide some good test matrices.

Feel free to review those and all future commits.

**Comment by c1887** [comment]
@nbbl The check for Markov property has some undesired (?) behavior; if I enter the transition matrix as a np.matrix object rather than np.array, the np.sum() will give me a matrix object as well, so the np.array_eqal method will return me a False, even though the sums of the matrix rows are one.

And for that matter, the _rows_ should sum to 1, not the columns. So I'll edit the error message.
Edit: Nevermind what I wrote here about the np.sum to change, that was fine.

**Comment by PhilippsCat** [comment]
i rewrote the code of today and make some little changes (described in the picture):
![imag0255](https://cloud.githubusercontent.com/assets/12544858/8112330/6274e080-1067-11e5-904b-75fb18401c66.jpg)

**Comment by c1887** [comment]
I have pushed a first k-means implementation to my branch. It seems to work somehow reasonable (see plot below), but I think there might be still some flaws in it. I had test cases where it failed to converge (I think it was the case when there was actually _no_ data point associated to a certain cluster). 
In the plot, the bigger circles are the estimated cluster centers and the colors of the circles describe the associated cluster.

![k_means_first_try](https://cloud.githubusercontent.com/assets/7395967/8235312/7d9d7d7a-15e1-11e5-8abc-05134a0d89f8.png)

**Comment by nwulkow** [comment]
Hi, I pushed two little methods which we might or might not use into my branch. Both methods are meant to give the user a 'feeling' for his data. One of the methods takes a chain and plots the occurrences of each state descendingly (so for example if the plot looks like a constant line all states are reached about equally). The other one computes the variances of the lines of the transition matrix and ranks the states by them, so one can see which states are the most 'reliable' as high variance for a state i suggests that only few states are likely to be reached

**Comment by c1887** [comment]
@nwulkow: Thank you for implementing the regspace method. Anyway, there might be room for improvement.
First of all, please check your indentings. Your method should be on the same level as the kmeans method. Right now, it is an sub method of it. Maybe it would be good to test your code in the python command line (not only in an ipython notebook) to avoid these little things.

After that, I still have problems testing your code. It seems to depend highly on the right dimensions of the input data, as I get (on the first glance) reasonable results but that are always the same even for different dmin. 
Looking at your code, I also don't understand why you check for _<_ dmin. I mean don't we create a new cluster if it's too far _away_ from the others?

And in the end, one thing that would be nice is if we could also get a discrete trajectory from it like in the kmeans method, I mean something with has only integer states in it (e.g. [0,4,2,1,3,3,4]...), so we can put it directly into the estimation methods.

Edit: sorry for closing, that was by accident.

**Comment by nwulkow** [comment]
Hi, when testing the MSM-methods today I ran into a problem using the hitting_prob-function. If I give it either a numpy-array or a list as an argument, it spits the following error: "The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()" in the line "if coefficients[i] == [0] \* numberOfStates:". Did I do something wrong or do we have to modfy the code there?

**Comment by PhilippsCat** [comment]
i noticed that in the master branch is an old version of hitting_prob.
the actual version is in my branch.
how can i merge it?

**Comment by nwulkow** [comment]
Just copy and paste it I would say

**Comment by c1887** [comment]
No. Use git commands as before. Johannes wrote at <del>some point</del> ([here](https://github.com/PhilippsCat/MarkovChains/issues/3))  how you do it in detail. In general, you update your local master branch ('git pull'), switch to your branch, perform a merge (`git merge master`). This way, your own branch should be up to date plus your personal changes (for ex. this new version of hitting_prob you mentioned), if there wasn't any errors of course. Then you can switch back to master branch, perform `git merge philippbranch` and if that doesn't raise any error, you can push it in the usual way.

**Comment by nwulkow** [comment]
If I try the version from your branch it still spits out the same error. If I have 'coefficients[i]' and '[0]_numberOfStates' printed, 'coefficients[i]' is an array that is not comma-seperated whereas [0]_numberOfStates is.

**Comment by nwulkow** [comment]
Maybe we could start talking in German here

**Comment by c1887** [comment]
Zu deinem Problem mit den Hitting_probs kann ich nichts sagen, aber @nwulkow funktioniert deine Regspace Implementierung? Weil dann würde ich die auch in master mit hereinnehmen.

**Comment by nwulkow** [comment]
Ja, ich habe jetzt 1000 verschiedene Tests damit gemacht und die haben funktioniert.

**Comment by c1887** [comment]
Gut.

Hab mal rudimentäre Unterstützung für PCCA eingebaut. Aber das basiert, wie wir uns ja schon überlegt hatten, darauf, dass man PyEMMA lokal installiert hat.

**Comment by PhilippsCat** [comment]
ich glaube ich sehe wo der fehler im code ist und werde das später ausbessern (bin grade auf arbeit).
liegt wohl daran, dass ich den code zu hause erst ohne numpy geschrieben habe und numpys matrixeigenschaften für den zugriff anders als gedacht sind. sollte so etwa 18 uhr funktionieren dann

**Comment by PhilippsCat** [comment]
ich bin noch auf arbeit heute, aber wenn du die zeile ersetzt durch:
if np.all(coefficients[i] == [0] \* numberOfStates): ?

**Comment by nwulkow** [comment]
Ja, jetzt läuft es zwar durch aber als Lösung kommt bei mir immer [1,1,1] raus für die Übergangsmatrix  T = np.array([[0.3,0.4,0.3],[0.4,0.3,0.3],[0.3,0.3,0.4]]) und für alle verschiedenen Mengen A, obwohl diese Bedingung nicht mal erfüllt ist und er dann auch nicht weiter reingeht

**Comment by PhilippsCat** [comment]
ja. 1,1,1 stimmt doch ?!
die hitting probability sagt doch aus wie wahrscheinlich es ist die menge a irgendwann zu treffen wenn ich den prozess unendlich lange laufen lasse und das konvergiert doch gegen 1 (wenn alle zustände miteinander kommunizieren)

**Comment by PhilippsCat** [comment]
di 12:00 computerraum
präsentation: do 15 uhr raum 126
